File: firstlast.py
#time complexity = O(lon(n))
#Space complexity = O(1)
#Ran successfully on leetcode
#no problems faced 

class Solution:
    
    def firstBinSearch(self,nums: List[int], target: int) -> int:
        low = 0
        high = len(nums) - 1
            
        while(low <= high):
            mid = low + ((high-low) // 2)
            if(nums[mid] == target):
                if(mid == 0 or nums[mid] > nums[mid-1]):
                    return mid
                else:
                    high = mid - 1
            elif(nums[mid] < target):
                low = mid + 1
            else:
                high = mid - 1
                    
        return -1
        
    def secondBinSearch(self,nums: List[int],target: int,low: int,high: int) -> int:
        
        while(low <= high):
            mid = low + (high - low) // 2
            if(nums[mid] == target):
                if(mid == len(nums) - 1 or nums[mid+1] > target):
                    return mid
                else:
                    low = mid + 1
            elif(nums[mid] > target):
                high = mid-1
            else:
                low = mid+1
        return -1
    
    
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        
        if(len(nums) == 0):
            return [-1,-1]
        
        if(nums[0] > target or nums[len(nums)-1] < target):
            return [-1,-1]
        
        first = self.firstBinSearch(nums,target)
        last = self.secondBinSearch(nums,target,first,len(nums)-1)
        return [first,last]

File: first-last.java
Time Complexity: O(log n)
Space Complexity: O(1)

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int first = binarySearch(nums, target, false);
        int last = binarySearch(nums, target, true);
        return new int[]{first, last};
    }
    
    private int binarySearch(int[] nums, int target, boolean last) {
        int low = 0;
        int high = nums.length - 1;
        while(low <= high) {
            int mid = low + (high - low)/2;
            if(nums[mid] == target) {
                //search for first
                if(!last) {
                    if(mid == 0 || nums[mid] > nums[mid-1]) {
                        return mid;
                    }else {
                        high = mid - 1;
                    }
                }else {
                    if(mid == nums.length - 1 || nums[mid] < nums[mid+1]) {
                        return mid;
                    }else {
                        low = mid + 1;
                    }
                }
                
                //search for last
                
            }else if(nums[mid] > target) {
                high = mid - 1;
            }else {
                low = mid + 1;
            }
        }
        return -1;
    }
}